File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/AppDetailDataService.java
package sk.styk.martin.apkanalyzer.business.service;

import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import sk.styk.martin.apkanalyzer.model.detail.AppDetailData;

/**
 * Retrieve apps installed on device
 * <p>
 * Created by Martin Styk on 14.06.2017.
 */
public class AppDetailDataService {

    private static final String TAG = AppDetailDataService.class.getSimpleName();

    private final int ANALYSIS_FLAGS = PackageManager.GET_SIGNATURES |
            PackageManager.GET_ACTIVITIES |
            PackageManager.GET_SERVICES |
            PackageManager.GET_PROVIDERS |
            PackageManager.GET_RECEIVERS |
            PackageManager.GET_PERMISSIONS |
            PackageManager.GET_CONFIGURATIONS;

    private PackageManager packageManager;

    private GeneralDataService generalDataService;
    private CertificateService certificateService;
    private AppComponentsService appComponentsService;
    private PermissionsService permissionsService;
    private FeaturesService featuresService;
    private FileDataService fileDataService;
    private ResourceService resourceService;
    private DexService dexService;

    public AppDetailDataService(@NonNull PackageManager packageManager) {
        this.packageManager = packageManager;
        this.generalDataService = new GeneralDataService(packageManager);
        this.certificateService = new CertificateService();
        this.appComponentsService = new AppComponentsService(packageManager);
        this.permissionsService = new PermissionsService();
        this.featuresService = new FeaturesService();
        this.fileDataService = new FileDataService();
        this.resourceService = new ResourceService();
        this.dexService = new DexService();
    }

    @Nullable
    public AppDetailData get(String packageName, String pathToPackage) {
        PackageInfo packageInfo = null;
        AppDetailData data = null;
        // decide whether we analyze installed app or only apk file
        if (packageName != null && pathToPackage == null) {
            data = new AppDetailData(AppDetailData.AnalysisMode.INSTALLED_PACKAGE);

            try {
                packageInfo = packageManager.getPackageInfo(packageName, ANALYSIS_FLAGS);
            } catch (Exception e) {
                e.printStackTrace();
            }
        } else if (packageName == null && pathToPackage != null) {
            data = new AppDetailData(AppDetailData.AnalysisMode.APK_FILE);

            packageInfo = packageManager.getPackageArchiveInfo(pathToPackage, ANALYSIS_FLAGS);
            if (packageInfo != null) packageInfo.applicationInfo.sourceDir = pathToPackage;
        }

        if (packageInfo == null)
            return null;

        try {
            data.setGeneralData(generalDataService.get(packageInfo));
            data.setCertificateData(certificateService.get(packageInfo));
            data.setActivityData(appComponentsService.getActivities(packageInfo));
            data.setServiceData(appComponentsService.getServices(packageInfo));
            data.setContentProviderData(appComponentsService.getContentProviders(packageInfo));
            data.setBroadcastReceiverData(appComponentsService.getBroadcastReceivers(packageInfo));
            data.setPermissionData(permissionsService.get(packageInfo));
            data.setFeatureData(featuresService.get(packageInfo));
            data.setFileData(fileDataService.get(packageInfo));
            data.setResourceData(resourceService.get(data.getFileData()));
            data.setClassPathData(dexService.get(packageInfo));
        } catch (Exception e) {
            // we catch a general exception here, because some repacakged APKs are really naughty
            // and we rather show user error screen than app failure
            Log.e(TAG, e.toString());
            return null;
        }
        return data;
    }

    @Override
    public String toString() {
        return "AppDetailDataService{" +
                "ANALYSIS_FLAGS=" + ANALYSIS_FLAGS +
                ", packageManager=" + packageManager +
                ", generalDataService=" + generalDataService +
                ", certificateService=" + certificateService +
                ", appComponentsService=" + appComponentsService +
                ", permissionsService=" + permissionsService +
                ", featuresService=" + featuresService +
                ", fileDataService=" + fileDataService +
                ", resourceService=" + resourceService +
                ", dexService=" + dexService +
                '}';
    }
}

File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/launcher/AppDetailDataService.kt
package sk.styk.martin.apkanalyzer.business.service.launcher

import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.util.Log
import sk.styk.martin.apkanalyzer.business.service.AppComponentsService
import sk.styk.martin.apkanalyzer.business.service.CertificateService
import sk.styk.martin.apkanalyzer.business.service.DexService
import sk.styk.martin.apkanalyzer.business.service.FeaturesService
import sk.styk.martin.apkanalyzer.business.service.FileDataService
import sk.styk.martin.apkanalyzer.business.service.GeneralDataService
import sk.styk.martin.apkanalyzer.business.service.PermissionsService
import sk.styk.martin.apkanalyzer.business.service.ResourceService
import sk.styk.martin.apkanalyzer.model.detail.AppDetailData

/**
 * Retrieve apps installed on device
 *
 * @author Martin Styk
 * @version 14.06.2017.
 */
class AppDetailDataService(private val packageManager: PackageManager) {
    private val TAG = AppDetailDataService::class.java.simpleName

    private val analysisFlags = PackageManager.GET_SIGNATURES or
            PackageManager.GET_ACTIVITIES or
            PackageManager.GET_SERVICES or
            PackageManager.GET_PROVIDERS or
            PackageManager.GET_RECEIVERS or
            PackageManager.GET_PERMISSIONS or
            PackageManager.GET_CONFIGURATIONS

    private val generalDataService = GeneralDataService()
    private val certificateService = CertificateService()
    private val appComponentsService = AppComponentsService()
    private val permissionsService = PermissionsService()
    private val featuresService = FeaturesService()
    private val fileDataService = FileDataService()
    private val resourceService = ResourceService()
    private val dexService = DexService()

    fun get(packageName: String?, pathToPackage: String?): AppDetailData? {

        val packageInfo: PackageInfo
        val analysisMode: AppDetailData.AnalysisMode?
        try {
            // decide whether we analyze installed app or only apk file
            packageInfo = when {
                !packageName.isNullOrBlank() && pathToPackage.isNullOrBlank() -> {
                    analysisMode = AppDetailData.AnalysisMode.INSTALLED_PACKAGE
                    packageManager.getPackageInfo(packageName, analysisFlags)
                }
                packageName.isNullOrBlank() && !pathToPackage.isNullOrBlank() -> {
                    analysisMode = AppDetailData.AnalysisMode.APK_FILE
                    packageManager.getPackageArchiveInfoWithCorrectPath(pathToPackage!!, analysisFlags)
                }
                else -> throw IllegalArgumentException("At least one way to get package needs to be specified  [$packageName/$pathToPackage]")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Can not read package info of package [$packageName/$pathToPackage] :$e")
            return null
        }

        return try {
            val fileData = fileDataService.get(packageInfo)

            AppDetailData(
                    analysisMode = analysisMode,
                    generalData = generalDataService.get(packageInfo, packageManager),
                    certificateData = certificateService.get(packageInfo),
                    activityData = appComponentsService.getActivities(packageInfo, packageManager),
                    serviceData = appComponentsService.getServices(packageInfo),
                    contentProviderData = appComponentsService.getContentProviders(packageInfo),
                    broadcastReceiverData = appComponentsService.getBroadcastReceivers(packageInfo),
                    permissionData = permissionsService.get(packageInfo, packageManager),
                    featureData = featuresService.get(packageInfo),
                    fileData = fileData,
                    resourceData = resourceService.get(fileData),
                    classPathData = dexService.get(packageInfo)
            )

        } catch (e: Exception) {
            // we catch a general exception here, because some repackaged APKs are really naughty
            // and we rather show user error screen than app failure
            Log.e(TAG, e.toString())
            null

        }

    }

    private fun PackageManager.getPackageArchiveInfoWithCorrectPath(pathToPackage: String, analysisFlags: Int): PackageInfo {
        val packageInfo = getPackageArchiveInfo(pathToPackage, analysisFlags)
        packageInfo?.applicationInfo?.sourceDir = pathToPackage

        return packageInfo
    }
}

